alternator/streams.cc

```diff
@@ -182,5 +182,277 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
     return make_ready_future<executor::request_return_type>(make_jsonable(std::move(ret)));
 }
 
+struct shard_id {
+    static constexpr auto marker = 'H';
+
+    utils::UUID table;
+    db_clock::time_point time;
+    cdc::stream_id id;
+
+    shard_id(utils::UUID tab, db_clock::time_point t, cdc::stream_id i)
+        : table(tab)
+        , time(t)
+        , id(i)
+    {}
+    shard_id(const sstring&);
+
+    friend std::ostream& operator<<(std::ostream& os, const shard_id& id) {
+        return os << marker << id.table << std::hex
+            << ':' << id.time.time_since_epoch().count()
+            << ':' << id.id.to_bytes()
+            ;
+    }
+};
+
+shard_id::shard_id(const sstring& s) {
+    auto i = s.find(':');
+    auto j = s.find(':', i + 1);
+
+    if (s.at(0) != marker || i == sstring::npos || j == sstring::npos) {
+        throw std::invalid_argument(s);
+    }
+
+    table = utils::UUID(s.substr(1, i));
+    time = db_clock::time_point(db_clock::duration(std::stoull(s.substr(i + 1, j), nullptr, 16)));
+    id = cdc::stream_id(from_hex(s.substr(j + 1)));
+}
+
+}
+
+template<typename ValueType>
+struct rapidjson::internal::TypeHelper<ValueType, alternator::shard_id>
+    : public from_string_helper<ValueType, alternator::shard_id>
+{};
+
+namespace alternator {
+
+enum class stream_view_type {
+    KEYS_ONLY,
+    NEW_IMAGE,
+    OLD_IMAGE,
+    NEW_AND_OLD_IMAGES,
+};
+
+std::ostream& operator<<(std::ostream& os, stream_view_type type) {
+    switch (type) {
+        case stream_view_type::KEYS_ONLY: os << "KEYS_ONLY"; break;
+        case stream_view_type::NEW_IMAGE: os << "NEW_IMAGE"; break;
+        case stream_view_type::OLD_IMAGE: os << "OLD_IMAGE"; break;
+        case stream_view_type::NEW_AND_OLD_IMAGES: os << "NEW_AND_OLD_IMAGES"; break;
+        default: throw std::invalid_argument(std::to_string(int(type)));
+    }
+    return os;
+}
+std::istream& operator>>(std::istream& is, stream_view_type& type) {
+    sstring s;
+    is >> s;
+    if (s == "KEYS_ONLY") {
+        type = stream_view_type::KEYS_ONLY;
+    } else if (s == "NEW_IMAGE") {
+        type = stream_view_type::NEW_IMAGE;
+    } else if (s == "OLD_IMAGE") {
+        type = stream_view_type::OLD_IMAGE;
+    } else if (s == "NEW_AND_OLD_IMAGES") {
+        type = stream_view_type::NEW_AND_OLD_IMAGES;
+    } else {
+        throw std::invalid_argument(s);
+    }
+    return is;
+}
+
+static stream_view_type cdc_options_to_steam_view_type(const cdc::options& opts) {
+    stream_view_type type = stream_view_type::KEYS_ONLY;
+    if (opts.preimage() && opts.postimage()) {
+        type = stream_view_type::NEW_AND_OLD_IMAGES;
+    } else if (opts.preimage()) {
+        type = stream_view_type::OLD_IMAGE;
```

Comment:
CDC itself only calculates postimage (from pre + delta) if postimage is enabled.

Reply:
Maybe CDC only calculates it, but isn't storing both on disk (in the CDC log) a waste of disk space? Wouldn't it be better to read the pre and delta and then *calculate*, on the fly, the post?
I don't mind that you add this as a "TODO" or "idea for optimization" for later, but I want to understand this...

Reply:
While alternator has fairly simplistic column (i.e. no non-atomics), in general, the merge pre-image + delta is not 100% trivial. Note that we also allow post-image without pre-image, and so does dynamo streams. 
Calculating this on-the-fly has some benefits, but it would at the very least force us to make the alternator stream config persisted instead of implicit via CDC settings.

Reply:
Very well. Let's take it like this, but I will expect you to either remember this issue, or (better) create an issue in the bug tracker about the inefficiency in the current implementation saving all of preimage, delta and postimage - all on disk.